omgnetwork/elixir-omg
# feat: add metrics: balance per currency, number of unspent outputs, number of addresses

Close #1148

## Overview

Add metrics for childchain's balances, number of unspent outputs and addresses with unspent outputs.

![image](https://user-images.githubusercontent.com/921194/88316360-e9f5f480-cd41-11ea-98b1-fc1522422c06.png)

## Changes

- Adds `total_unspent_outputs`: The total number of unspent outputs being tracked in the childchain service
- Adds `total_unspent_addresses`: The total number of addresses holding at least 1 unspent output
- Updates `balance`: Same definition. Fixed where the balance was calculating only from the recent cache
- Removes `unique_users`: Replaced by `total_unspent_addresses`

Note that I originally wanted to avoid or optimize utxos calling but it requires lower level opimization and the current DB can handle the load for the foreseeable future (see https://github.com/omgnetwork/elixir-omg/pull/1667 for some benchmark). 

With this PR we get the utxos db observability and can monitor when the metrics become concerning so I think it's worth the extra call.

## Testing

Follow these steps:

1. Configure `DD_DISABLED: false` and `DD_API_KEY` at least on the childchain
1. Run `make docker-build`
1. Run `cd priv/cabbage && make start_daemon_services-2`
1. Make some deposits and transfers
1. The 3 metrics should be reported
